### PR TITLE
Feature/Boost Logger Implementation

### PR DESCRIPTION
**Pull request by ShobhitAd**
This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Summary
Implements a third party logger for sdl core

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/utils/src/logger/boostlogger.cc [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+  }
+}
+
+boost::posix_time::ptime BoostLogger::GetLocalPosixTime(
+    const logger::TimePoint& timestamp) {
+  auto time = std::chrono::duration_cast<std::chrono::microseconds>(
+                  timestamp.time_since_epoch())
+                  .count();
+
+  boost::posix_time::ptime time_epoch(boost::gregorian::date(1970, 1, 1));
```
Comment (iCollin) [inline src/components/utils/src/logger/boostlogger.cc:145]: can we make this constexpr? is this not already a compile time constant in the boost lib?

File: src/components/utils/src/logger/boostlogger.cc [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    auto& sinks = settings.property_tree().get_child("Sinks");
+    for (auto& sink : sinks) {
+      std::string sink_name = "Sinks." + sink.first;
+
+      // Disable logging for particular sinks
+      if (boost::optional<std::string> param =
+              settings[sink_name]["DisableLogging"]) {
+        bool sink_disabled = (param.get() == "true");
+        if (sink_disabled) {
+          sinks.erase(sink.first);
```
Comment (iCollin) [inline src/components/utils/src/logger/boostlogger.cc:107]: calling erase invalidates the iterator we are looping on

File: src/components/telemetry_monitor/test/CMakeLists.txt [hunk 3]
```diff
@@ -74,4 +74,10 @@ endif()
 
 create_test("telemetry_monitor_test" "${SOURCES}" "${LIBRARIES}")
 
-file(COPY ${COMPONENTS_DIR}/telemetry_monitor/test/log4cxx.properties DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
+if (${LOGGER_NAME} STREQUAL "LOG4CXX")
+  file(COPY ${COMPONENTS_DIR}/telemetry_monitor/test/log4cxx.properties DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
+endif()
+
```
Comment (iCollin) [inline src/components/telemetry_monitor/test/CMakeLists.txt:80]: can this be else?

File: src/appMain/boostlogconfig.ini [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+DisableLogging=false
+Destination=TextFile
+FileName=TransportManager.log
+Append=false
+Filter="%Component% = TransportManager"
+Format="%Severity% [%TimeStamp%][%ThreadId%][%Component%] %FileName%:%LineNum% %Trace%: %Message%"
+AutoFlush=true
+Asynchronous=true
+
+# Log file for handling Ford protocol info (include ProtocolHandler, ConnectionHandler, SecurityManager)
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply src/appMain/boostlogconfig.ini:35]: Just trying to match the same configuration that exists https://github.com/smartdevicelink/sdl_core/blob/develop/src/appMain/log4cxx.properties#L37. Not particularly sure if that log file is commonly used but if we change it I think it should be changed for both boost and log4cxx.

File: src/components/test_main.cc [hunk 5]
```diff
@@ -15,8 +14,6 @@ int main(int argc, char** argv) {
   // --------------------------------------------------------------------------
   // Logger initialization
   // Redefine for each paticular logger implementation
-  auto logger = std::unique_ptr<logger::Log4CXXLogger>(
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply src/components/test_main.cc:18]: That variable doesn't seem to be used anywhere. The logger_impl in the next line is initialized with `new logger::LoggerImpl(false)`

File: CMakeLists.txt [hunk 6]
```diff
@@ -59,6 +59,12 @@ option(USE_GOLD_LD "Use gold linker intead of GNU linker" ON)
 option(USE_CCACHE "Turn on ccache usage" ON)
 option(USE_DISTCC "Turn on distributed build_usage" OFF)
 
+set(LOGGER_NAME "BOOST" CACHE STRING "Logging library to use (BOOST, LOG4CXX)")
+set_property(CACHE LOGGER_NAME PROPERTY STRINGS BOOST LOG4CXX)
+if(LOGGER_NAME STREQUAL "")
+  set(LOGGER_NAME "BOOST")
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply CMakeLists.txt]: Changed in https://github.com/smartdevicelink/sdl_core/pull/3571/commits/31860450ad5362ae084e02152f5b14f47988ed7a

File: src/appMain/CMakeLists.txt [hunk 7]
```diff
@@ -160,7 +160,13 @@ target_link_libraries(${PROJECT} ${LIBRARIES})
 
 add_dependencies(${PROJECT} Policy)
 file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/build_config.txt DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
-file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/log4cxx.properties DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
+if (${LOGGER_NAME} STREQUAL "LOG4CXX")
+  file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/log4cxx.properties DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
+endif()
+if (${LOGGER_NAME} STREQUAL "BOOST")
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply src/appMain/CMakeLists.txt]: Changed in https://github.com/smartdevicelink/sdl_core/pull/3571/commits/31860450ad5362ae084e02152f5b14f47988ed7a

File: tools/policy_table_validator/main.cpp [hunk 8]
```diff
@@ -55,10 +60,16 @@ int main(int argc, char** argv) {
   // --------------------------------------------------------------------------
   // Logger initialization
   // Redefine for each paticular logger implementation
-  auto logger = std::unique_ptr<logger::Log4CXXLogger>(new logger::Log4CXXLogger("log4cxx.properties"));
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply tools/policy_table_validator/main.cpp:58]: My apologies, I commented it out while testing out something and forgot to uncomment it. Fixed in https://github.com/smartdevicelink/sdl_core/pull/3571/commits/31860450ad5362ae084e02152f5b14f47988ed7a.

just fyi, it was causing an `undefined reference to 'boost:basic_regex ...'` error but adding https://github.com/smartdevicelink/sdl_core/pull/3571/files#diff-7da77e2dbbe71261984dccc087ea8f3efedda09a22bae9d934a9de5270869939R132 seems to fix it.

File: src/components/utils/test/CMakeLists.txt [hunk 9]
```diff
@@ -93,7 +93,13 @@ set(LIBRARIES
 create_test(utils_test "${SOURCES}" "${LIBRARIES}")
 
 file(COPY testscript.sh DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
-file(COPY log4cxx.properties DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
+if (${LOGGER_NAME} STREQUAL "LOG4CXX")
+  file(COPY log4cxx.properties DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
+endif()
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply src/components/utils/test/CMakeLists.txt]: Changed in https://github.com/smartdevicelink/sdl_core/pull/3571/commits/31860450ad5362ae084e02152f5b14f47988ed7a

File: src/components/telemetry_monitor/test/CMakeLists.txt (hunk 3 above)
Comment (ShobhitAd) [reply src/components/telemetry_monitor/test/CMakeLists.txt:80]: Changed in https://github.com/smartdevicelink/sdl_core/pull/3571/commits/31860450ad5362ae084e02152f5b14f47988ed7a

File: src/components/utils/src/logger/boostlogger.cc (hunk 1 above)
Comment (ShobhitAd) [reply src/components/utils/src/logger/boostlogger.cc:145]: Wasn't able to find a specific constant for the epoch(January 1, 1970). Also not sure if `boost::gregorian::date` meets the constexpr Literal type requirements https://en.cppreference.com/w/cpp/named_req/LiteralType. I could make a separate constant for the `epoch_date`. Let me know what you think

File: src/3rd_party/CMakeLists.txt [hunk 10]
```diff
@@ -221,7 +221,8 @@ else()
 endif()
 
 set(BOOST_ROOT ${3RD_PARTY_INSTALL_PREFIX})
-find_package(Boost 1.66.0 COMPONENTS system thread date_time filesystem regex) 
+find_package(Boost 1.66.0 COMPONENTS system thread date_time filesystem regex log log_setup)
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply src/3rd_party/CMakeLists.txt]: Fixed in https://github.com/smartdevicelink/sdl_core/pull/3571/commits/f7f7ac3ddc14c887c101cbbf054fde72df721580

File: src/components/utils/CMakeLists.txt [hunk 11]
```diff
@@ -121,6 +127,8 @@ list(APPEND LIBRARIES boost_system -L${BOOST_LIBS_DIRECTORY})
 list(APPEND LIBRARIES boost_thread -L${BOOST_LIBS_DIRECTORY})
 list(APPEND LIBRARIES boost_date_time -L${BOOST_LIBS_DIRECTORY})
 list(APPEND LIBRARIES boost_filesystem -L${BOOST_LIBS_DIRECTORY})
+list(APPEND LIBRARIES boost_log_setup -L${BOOST_LIBS_DIRECTORY})
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply src/components/utils/CMakeLists.txt]: Fixed in https://github.com/smartdevicelink/sdl_core/pull/3571/commits/f7f7ac3ddc14c887c101cbbf054fde72df721580

File: src/components/utils/CMakeLists.txt [hunk 12]
```diff
@@ -127,14 +127,21 @@ list(APPEND LIBRARIES boost_system -L${BOOST_LIBS_DIRECTORY})
 list(APPEND LIBRARIES boost_thread -L${BOOST_LIBS_DIRECTORY})
 list(APPEND LIBRARIES boost_date_time -L${BOOST_LIBS_DIRECTORY})
 list(APPEND LIBRARIES boost_filesystem -L${BOOST_LIBS_DIRECTORY})
-list(APPEND LIBRARIES boost_log_setup -L${BOOST_LIBS_DIRECTORY})
-list(APPEND LIBRARIES boost_log -L${BOOST_LIBS_DIRECTORY})
+if (ENABLE_LOG AND ${LOGGER_NAME} STREQUAL "BOOST")
+  list(APPEND LIBRARIES boost_log_setup -L${BOOST_LIBS_DIRECTORY})
+  list(APPEND LIBRARIES boost_log -L${BOOST_LIBS_DIRECTORY})
+endif ()
+list(APPEND LIBRARIES boost_regex -L${BOOST_LIBS_DIRECTORY})
```
Comment (iCollin) [inline src/components/utils/CMakeLists.txt:134]: do you know why this worked without this append?

File: src/appMain/boostlogconfig.ini (hunk 4 above)
[the remark this answers is not in the document]
Comment (iCollin) [reply src/appMain/boostlogconfig.ini:35]: that's fair, it's good for things to match.

File: src/components/utils/src/logger/boostlogger.cc (hunk 1 above)
Comment (iCollin) [reply src/components/utils/src/logger/boostlogger.cc:145]: yup, i couldn't find a constant either. seems the way you're doing this is common :)

File: src/components/utils/src/logger/boostlogger.cc (hunk 2 above)
Comment (ShobhitAd) [reply src/components/utils/src/logger/boostlogger.cc]: Fixed in https://github.com/smartdevicelink/sdl_core/pull/3571/commits/17a4d07d9c19313ad2bebda119e69f5825180956

File: src/components/utils/src/logger/boostlogger.cc [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+    return std::string(full_function_signature);
+  }
+
+  // Get the function name(including namespaces) from the function signature
+  return std::string(results[1]);
+}
+
+bool BoostLogger::IsEnabledFor(const std::string& component,
+                               LogLevel log_level) const {
+  // To be implemented once the config file adds limiting the log level
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply src/components/utils/src/logger/boostlogger.cc]: Updated the note in https://github.com/smartdevicelink/sdl_core/pull/3571/commits/7cdb2c50c4b9bf563b4371f9188070f8699cc63f. Currently, basic filtering based on log_level and component name is handled using the value in the `Filter` attribute. 
This function should be implemented if we end up adding more custom attributes for filtering or require additional logic to filter based on log_level and component name.

**Review by iCollin — COMMENTED** [review]
Is sudo necessary for the boost install?

File: src/components/utils/src/logger/boostlogger.cc [hunk 14]
```diff
@@ -96,17 +96,25 @@ void BoostLogger::Init() {
...
-      std::string sink_name = "Sinks." + sink.first;
+
+    for (auto it = sinks.begin(); it != sinks.end();) {
+      bool to_be_removed = false;
+      std::string sink_name = "Sinks." + it->first;
 
       // Disable logging for particular sinks
       if (boost::optional<std::string> param =
               settings[sink_name]["DisableLogging"]) {
         bool sink_disabled = (param.get() == "true");
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply src/components/utils/src/logger/boostlogger.cc]: Makes sense. Changed in https://github.com/smartdevicelink/sdl_core/pull/3571/commits/40184cc887cc309161453432dd964dbd74e260d1

**Comment by ShobhitAd** [comment]
> Is sudo necessary for the boost install?

I don't think `sudo` is necessary to compile and install the boost components but since the `3RD_PARTY_INSTALL_PREFIX` is set to `/usr/local`(https://github.com/smartdevicelink/sdl_core/blob/master/src/3rd_party/set_3rd_party_paths.cmake#L50) we do require special permissions to write to that directory. 

The install target could be changed but it would end up affecting more than just the boost libraries(libson and log4cxx). That said it looks like `sudo` is only required for the first time setup(when the library is not found) all subsequent build will not require `sudo`.

**Review by iCollin — APPROVED** [review]

File: src/appMain/main.cc [hunk 15]
```diff
@@ -142,17 +149,25 @@ int32_t main(int32_t argc, char** argv) {
...
   }
 #endif
 
   threads::Thread::SetNameForId(threads::Thread::CurrentId(), "SDLCore");
 
+#ifdef LOG4CXX_LOGGER
   if (!utils::appenders_loader.Loaded()) {
     SDL_LOG_ERROR("Appenders plugin not loaded, file logging disabled");
   }
+#endif  // LOG4CXX_LOGGER
```
Comment (jacobkeeler) [inline src/appMain/main.cc:170]: Seems this section causes the build to fail if `ENABLE_LOG=OFF`, maybe move it into the `ENABLE_LOG` ifdef?

**Review by jacobkeeler — CHANGES_REQUESTED** [review]

File: src/appMain/main.cc (hunk 15 above)
Comment (ShobhitAd) [reply src/appMain/main.cc]: Resolved in https://github.com/smartdevicelink/sdl_core/pull/3571/commits/770ef067050eb4100ea18d03489fe017025f71f8

**Review by jacobkeeler — APPROVED** [review]

**Pull request merged**